apis/goharbor.io/v1beta1/harbor_configuration.go

```diff
@@ -0,0 +1,328 @@
+
+import (
+	"encoding/json"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Extension struct {
+	// Work around for https://github.com/kubernetes-sigs/kubebuilder/issues/528
+	Data map[string]interface{} `json:"-"`
+}
+
+// MarshalJSON marshals the HelmValues data to a JSON blob.
+func (e Extension) MarshalJSON() ([]byte, error) {
+	return json.Marshal(e.Data)
+}
+
+// UnmarshalJSON sets the HelmValues to a copy of data.
+func (e *Extension) UnmarshalJSON(data []byte) error {
+	var out map[string]interface{}
+	err := json.Unmarshal(data, &out)
+	if err != nil {
+		return err
+	}
+	e.Data = out
+	return nil
+}
+
+// DeepCopyInto is an deepcopy function, copying the receiver, writing
+// into out. In must be non-nil. Declaring this here prevents it from
+// being generated in `zz_generated.deepcopy.go`.
+//
+// This exists here to work around https://github.com/kubernetes/code-generator/issues/50,
+// and partially around https://github.com/kubernetes-sigs/controller-tools/pull/126
+// and https://github.com/kubernetes-sigs/controller-tools/issues/294.
+func (in *Extension) DeepCopyInto(out *Extension) {
+	b, err := json.Marshal(in.Data)
+	if err != nil {
+		// The marshal should have been performed cleanly as otherwise
+		// the resource would not have been created by the API server.
+		panic(err)
+	}
+	var c map[string]interface{}
+	err = json.Unmarshal(b, &c)
+	if err != nil {
+		panic(err)
+	}
+	out.Data = c
+}
+
+// +genclient
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// +kubebuilder:object:root=true
+// +kubebuilder:storageversion
+// +k8s:openapi-gen=true
+// +resource:path=harborconfiguration
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:categories="goharbor",shortName="hc"
+// +kubebuilder:printcolumn:name="HarborName",type=string,JSONPath=`.metadata.labels['harbor-name']`,description="Harbor instance name"
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=`.status.status`,description="HarborConfiguration status"
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`,description="Timestamp representing the server time when this object was created. It is not guaranteed to be set in happens-before order across separate operations. Clients may not set this value. It is represented in RFC3339 form and is in UTC."
+// HarborConfiguration is the Schema for the harbors configuration.
+type HarborConfiguration struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec HarborConfigurationSpec `json:"spec,omitempty"`
+
+	Status HarborConfigurationStatus `json:"status,omitempty"`
+}
+
+// HarborConfigurationSpec defines the spec of HarborConfiguration.
+type HarborConfigurationSpec struct {
+	// AdditionalProperties provides workaround for those unknown configuration fields in the future.
+	// +kubebuilder:validation:Optional
+	Extension `json:",inline"`
```

Comment:
If the user fills in auth_mode, is it recorded to the AuthMode field or this field?

Reply:
@chlins 

Any answers here?

Reply:
Extension will catch all field,  so maybe we cannot use the json inline format, extension should as external field. Code updated.